### PR TITLE
Gh 321 blockchain api

### PR DESCRIPTION
**Pull request by chb0github**
Initial stab at blockchain API

### CODE REVIEW

**Comment by prince-chrismc** [comment]
old PR comments https://github.com/chb0github/pyrsia/pull/4

File: src/blockchain/src/blockchain.rs [hunk 1]
```diff
@@ -94,8 +95,14 @@ impl GenesisBlock {
     }
 }
 
-#[derive(Serialize, Deserialize, Debug, Clone)]
+#[derive(Serialize, Deserialize)]
 pub struct Blockchain {
+    // this seems suspicious to chris
```
Comment (prince-chrismc) [inline src/blockchain/src/blockchain.rs]: Seems taken care of

```suggestion
```

File: src/blockchain/src/notifier.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
+use std::fmt;
+use std::fmt::{Debug, Formatter};
+
+use crate::block::{Block, Transaction};
+use crate::blockchain::Blockchain;
+
+#[derive(Hash, Eq, PartialEq, Clone)]
+pub struct BlockchainError;
```
Comment (prince-chrismc) [inline src/blockchain/src/notifier.rs]: I can not find this being returned anywhere 🤔 How is it being used?

File: src/blockchain/src/notifier.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+impl Debug for Blockchain {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let Blockchain {
+            trans_observers: _,
+            genesis_block,
+            blocks,
+            block_observers: _,
+        } = self;
+
+        f.debug_struct("Blockchain")
```
Comment (prince-chrismc) [inline src/blockchain/src/notifier.rs]: Maybe print the number of observers? give some clue

File: src/blockchain/src/notifier.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+use std::fmt::{Debug, Formatter};
+
+use crate::block::{Block, Transaction};
+use crate::blockchain::Blockchain;
+
+#[derive(Hash, Eq, PartialEq, Clone)]
+pub struct BlockchainError;
+
+// But we require certain bounds to get things done...
+impl Blockchain {
```
Comment (prince-chrismc) [inline src/blockchain/src/notifier.rs]: in Rust the `impl` of a type usually is in the same file.

It's strange to see this so far away... do you have any objections to moving this?

File: src/blockchain/src/blockchain.rs (hunk 1 above)
Comment (chb0github) [reply src/blockchain/src/blockchain.rs]: I'll remove the comment

File: src/blockchain/src/notifier.rs (hunk 2 above)
Comment (chb0github) [reply src/blockchain/src/notifier.rs]: It's not. I'll remove it

File: src/blockchain/src/notifier.rs (hunk 3 above)
Comment (chb0github) [reply src/blockchain/src/notifier.rs]: Fair.

File: src/blockchain/src/notifier.rs (hunk 4 above)
Comment (chb0github) [reply src/blockchain/src/notifier.rs]: No, none at all. I'll move it

**Review by prince-chrismc — APPROVED** [review]
LGTM. open against the main repo 👍

**Comment by chb0github** [comment]
Gh-321

**Pull request merged**